apiserver/facades/client/resources/repository.go

```diff
@@ -43,71 +43,177 @@ type CharmHub interface {
+	if !ok {
+		// This indicates that AddPendingResources() was called for
+		// a resource the charm backend doesn't know about (for the
+		// relevant charm revision).
+		return res, nil
+	}
+
+	if res.Revision < 0 {
+		// The caller wants to use the charm backend info.
+		return storeRes, nil
+	}
+	if res.Revision == storeRes.Revision {
+		// We don't worry about if they otherwise match. Only the
+		// revision is significant here. So we use the info from the
+		// charm backend since it is authoritative.
+		return storeRes, nil
+	}
+	if res.Fingerprint.IsZero() {
+		// The caller wants resource info from the charm backend, but with
+		// a different resource revision than the one associated with
+		// the charm in the backend.
+		return c.client.ResourceInfo(c.id.URL, c.id.Origin, res.Name, res.Revision)
+	}
+	// The caller fully-specified a resource with a different resource
+	// revision than the one associated with the charm in the backend. So
+	// we use the provided info as-is.
+	return res, nil
+}
+
 type charmHubClient struct {
+	resourceClient
 	client CharmHub
-	id     CharmID
 }
 
-// ResolveResources, looks at the provided, charmhub and backend (already downloaded)
-// resources to determine which to use.  Provided (uploaded) take precedence.  If
-// charmhub has a newer resource than the back end, use that.
-// TODO: (hml) 2020-12-03
-// this logic will need refinement as work to incorporate charmhub resources continues.
-// Right now, just take the uploaded resources and add charmhub resources where resource
-// is missing
-func (ch *charmHubClient) ResolveResources(uploadedResources []charmresource.Resource) ([]charmresource.Resource, error) {
-	chResources, err := ch.listResources()
+func newCharmHubClient(client CharmHub, id CharmID) *charmHubClient {
+	c := &charmHubClient{
+		client: client,
+	}
+	c.resourceClient = resourceClient{
+		client: c,
+		id:     id,
+	}
+	return c
+}
+
+// ResolveResources, looks at the provided, charmhub and backend (already
+// downloaded) resources to determine which to use. Provided (uploaded) take
+// precedence. If charmhub has a newer resource than the back end, use that.
+func (ch *charmHubClient) ResolveResources(resources []charmresource.Resource) ([]charmresource.Resource, error) {
+	storeResources, err := ch.listResources()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	for _, res := range uploadedResources {
-		if res.Origin == charmresource.OriginUpload {
-			chResources[res.Name] = res
-			continue
+	resolved, err := ch.resolveResources(resources, storeResources)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return resolved, nil
+}
+
+func (ch *charmHubClient) ResourceInfo(curl *charm.URL, origin corecharm.Origin, name string, revision int) (charmresource.Resource, error) {
+	if origin.ID == "" {
+		return charmresource.Resource{}, errors.Errorf("empty charm ID")
+	}
+
+	// We prefer the revision over the channel, so we attempt to locate the
```

Comment:
Why do we prefer revision over channel here?  With the charms, its the other way around?

For resources in charmstore, we use the newest revision of the resource, even if it's newer than what's in the charm revision.

Reply:
This is because we know exactly which charm we have, so we want to locate the exact resource revision associated (match exactly for the case there has been an update in between deploying a charm and deploying a resource). This is the opposite when selecting a charm, we don't know what revision to look for, so use the channel (default release).

If the revision is found then fall back to the latest channel.